### ISSUE TITLE
Missing events.

### ISSUE THREAD

**Issue by simonjohansson** [cloudfoundry/noaa#8]
Howdie!

Im using https://github.com/cloudfoundry-samples/fib-cpu in a newly deployed empty CF 211 VMWare installation. For each request to the app we get two log events

```
$ curl fib.test.cf.domain.com/10
$ cf logs fib
...
2015-06-23T15:26:40.28+0200 [App/0]      OUT Request /10 took 3.3326e-05 seconds to complete
2015-06-23T15:26:40.29+0200 [RTR/0]      OUT fib.test.cf.domain.com - [23/06/2015:13:26:40 +0000] "GET /10 HTTP/1.1" 200 0 2 "-" "curl/7.43.0" 10.230.31.7:51182 x_forwarded_for:"10.230.31.7" vcap_request_id:ce6fb1e7-ff26-497b-6c5c-b11cd47fb70d response_time:0.009590970 app_id:082fea25-f40b-4bff-8c12-7586447cb6c4
```

Using the noaa-client I would expect that shooting off 10 requests would yield 20 events but I get between 10-18 for each test I do.

Setup
```
$ GOPATH=/tmp go get github.com/cloudfoundry/noaa
$ cd /tmp/src/github.com/cloudfoundry/noaa/
$ git --no-pager show --summary 
commit 7f3b03afe00e1bf449cf1651a8623d878e20c70d
Author: John Tuley <jtuley@pivotal.io>
Date:   Thu Jun 18 16:50:33 2015 -0600

    Revert "Set gogo/protobuf version to be the same as the one in loggregator for travis builds"

    This reverts commit 3c78d4555842c8a7396fa4fc3ef18376c08a3e31.

    [#97345078]
```

Consuming and comparing events

```
term1$ cf logs fib > /tmp/logs-from-cf-logs
term2$ CF_ACCESS_TOKEN=wryyyy DOPPLER_ADDR=wss://doppler.test.cf.domain.com:443 GOPATH=/tmp go run firehose_sample/main.go > /tmp/noaa-client
term3$ repeat 10 curl fib.test.cf.domain.com/10 
term3$ cat /tmp/logs-from-cf-logs | egrep "App|RTR" | wc -l
20
term3$ cat /tmp/noaa-client | egrep "LogMessage" | wc -l  
17
$ cat /tmp/noaa-client | egrep "LogMessage"        
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066555822822483 logMessage:<message:"Request /10 took 3.3829e-05 seconds to complete" message_type:OUT timestamp:1435066555822820553 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"router__0" eventType:LogMessage timestamp:1435066555829558583 logMessage:<message:"fib.test.cf.domain.com - [23/06/2015:13:35:55 +0000] \"GET /10 HTTP/1.1\" 200 0 2 \"-\" \"curl/7.43.0\" 10.230.31.3:51264 x_forwarded_for:\"10.230.31.3\" vcap_request_id:37c3dd2f-49d4-49e5-4d85-52cf57baac84 response_time:0.007996051 app_id:082fea25-f40b-4bff-8c12-7586447cb6c4\n" message_type:OUT timestamp:1435066555829556378 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"RTR" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066555891428615 logMessage:<message:"Request /10 took 3.2354e-05 seconds to complete" message_type:OUT timestamp:1435066555891427679 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066555972095337 logMessage:<message:"Request /10 took 7.346e-05 seconds to complete" message_type:OUT timestamp:1435066555972093060 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"router__0" eventType:LogMessage timestamp:1435066555974757126 logMessage:<message:"fib.test.cf.domain.com - [23/06/2015:13:35:55 +0000] \"GET /10 HTTP/1.1\" 200 0 2 \"-\" \"curl/7.43.0\" 10.230.31.1:51266 x_forwarded_for:\"10.230.31.1\" vcap_request_id:5efaea2d-aee3-4c30-4e6d-91b354d5bf92 response_time:0.008960204 app_id:082fea25-f40b-4bff-8c12-7586447cb6c4\n" message_type:OUT timestamp:1435066555974755399 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"RTR" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066556038989559 logMessage:<message:"Request /10 took 5.659e-05 seconds to complete" message_type:OUT timestamp:1435066556038988623 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066556105148981 logMessage:<message:"Request /10 took 3.25e-05 seconds to complete" message_type:OUT timestamp:1435066556105136936 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"router__0" eventType:LogMessage timestamp:1435066556107187549 logMessage:<message:"fib.test.cf.domain.com - [23/06/2015:13:35:56 +0000] \"GET /10 HTTP/1.1\" 200 0 2 \"-\" \"curl/7.43.0\" 10.230.31.5:51268 x_forwarded_for:\"10.230.31.5\" vcap_request_id:06cf7d10-f3d7-47fb-6227-86eea2813bdd response_time:0.005772654 app_id:082fea25-f40b-4bff-8c12-7586447cb6c4\n" message_type:OUT timestamp:1435066556107185859 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"RTR" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066556187180248 logMessage:<message:"Request /10 took 7.2833e-05 seconds to complete" message_type:OUT timestamp:1435066556187178939 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"router__0" eventType:LogMessage timestamp:1435066556190834073 logMessage:<message:"fib.test.cf.domain.com - [23/06/2015:13:35:56 +0000] \"GET /10 HTTP/1.1\" 200 0 2 \"-\" \"curl/7.43.0\" 10.230.31.4:51269 x_forwarded_for:\"10.230.31.4\" vcap_request_id:fdf08d66-fe09-4578-7545-921207b5ca2e response_time:0.008445217 app_id:082fea25-f40b-4bff-8c12-7586447cb6c4\n" message_type:OUT timestamp:1435066556190832236 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"RTR" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066556273007272 logMessage:<message:"Request /10 took 3.2422e-05 seconds to complete" message_type:OUT timestamp:1435066556273005906 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"router__0" eventType:LogMessage timestamp:1435066556276260011 logMessage:<message:"fib.test.cf.domain.com - [23/06/2015:13:35:56 +0000] \"GET /10 HTTP/1.1\" 200 0 2 \"-\" \"curl/7.43.0\" 10.230.31.3:51270 x_forwarded_for:\"10.230.31.3\" vcap_request_id:86dfedaf-27aa-4333-6542-5f17171afe9b response_time:0.006513990 app_id:082fea25-f40b-4bff-8c12-7586447cb6c4\n" message_type:OUT timestamp:1435066556276258104 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"RTR" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066556336575706 logMessage:<message:"Request /10 took 3.2241e-05 seconds to complete" message_type:OUT timestamp:1435066556336574740 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066556441946000 logMessage:<message:"Request /10 took 7.332e-05 seconds to complete" message_type:OUT timestamp:1435066556441944975 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"router__0" eventType:LogMessage timestamp:1435066556445317709 logMessage:<message:"fib.test.cf.domain.com - [23/06/2015:13:35:56 +0000] \"GET /10 HTTP/1.1\" 200 0 2 \"-\" \"curl/7.43.0\" 10.230.31.2:51272 x_forwarded_for:\"10.230.31.2\" vcap_request_id:8ee423bd-88d7-4b02-5605-290e9935d259 response_time:0.008849451 app_id:082fea25-f40b-4bff-8c12-7586447cb6c4\n" message_type:OUT timestamp:1435066556445315729 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"RTR" source_instance:"0" >  
origin:"dea_logging_agent" eventType:LogMessage timestamp:1435066556529838323 logMessage:<message:"Request /10 took 7.2413e-05 seconds to complete" message_type:OUT timestamp:1435066556529837317 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"App" source_instance:"0" >  
origin:"router__0" eventType:LogMessage timestamp:1435066556533252926 logMessage:<message:"fib.test.cf.domain.com - [23/06/2015:13:35:56 +0000] \"GET /10 HTTP/1.1\" 200 0 2 \"-\" \"curl/7.43.0\" 10.230.31.6:51273 x_forwarded_for:\"10.230.31.6\" vcap_request_id:ac3e6844-3b6f-4214-6995-c888113b930d response_time:0.008822410 app_id:082fea25-f40b-4bff-8c12-7586447cb6c4\n" message_type:OUT timestamp:1435066556533250929 app_id:"082fea25-f40b-4bff-8c12-7586447cb6c4" source_type:"RTR" source_instance:"0" >
```

Let me know if you need any other info!

**Comment by simonjohansson** [comment]
Some updates.

On a standard CF 212 on VMWare, 2 doppler, 100 max messages in buffer.

sonde-go: cloudfoundry/sonde-go@e9b4698e850553401d1d2dd998c756fdf29356ea
noaa: 280497e4cd656fbf56aac4de5f0ad25f2d68758d

Running with cf-env that for each request logs two events, one from gorouter and one from the app itself.
Shooting off 1000 requests in quick succession I should see 2000 events.

With sample/main.go that only logs events for one app and the following patch
```
diff --git a/sample/main.go b/sample/main.go
index a8bcd02..2b937c9 100644
--- a/sample/main.go
+++ b/sample/main.go
@@ -17,17 +17,6 @@ func main() {
        connection := noaa.NewConsumer(dopplerAddress, &tls.Config{InsecureSkipVerify: true}, nil)
        connection.SetDebugPrinter(ConsoleDebugPrinter{})
 
-       messages, err := connection.RecentLogs(appGuid, authToken)
-
-       if err != nil {
-               fmt.Printf("===== Error getting recent messages: %v\n", err)
-       } else {
-               fmt.Println("===== Recent logs")
-               for _, msg := range messages {
-                       fmt.Println(msg)
-               }
-       }
-
        fmt.Println("===== Streaming metrics")
        msgChan := make(chan *events.Envelope)
        go func() {
```

I reliably get 2000 events every time.

With firehose_sample/main.go that logs all events I get between 500-1700.
I run each test 10 times, running more than one firehose_sample with the same subscription-id doesnt solve the issue.
 
Scaling doppler and buffer size etc does not help in my case.

TLDR
Consuming events for one app works correctly, when consuming all events and filter for the app events I want I only get a subset of the expected.

**Comment by jtuchscherer** [comment]
@simonjohansson Thanks for reporting this and investigating. And sorry that we didn't respond faster.
This issue is in our backlog and I will talk to @erikjasiak (our logging PM) to see if we can prioritize that among our upcoming work. 

Regarding your last comment, were the 500 - 1700 events all log messages from the app or are the other firehose events part of that number as well (in other words: did you get the 500 - 1700 before or after filtering)?

**Comment by jtuchscherer** [comment]
Also, could you please verify that you don't see any warning about the log output being to high?

You could just grep for the following string:
`Log message output too high.`

This message is generated here: https://github.com/cloudfoundry/loggregator/blob/develop/src/doppler/truncatingbuffer/truncating_buffer.go#L101

**Comment by simonjohansson** [comment]
Hi Johannes. There is lots more events besides the app logs, ie 500-1700
were filtered. Also I would say among the filtered events 75-80% are app
logs and the rest are rtr logs.

I'll provide better numbers tomorrow.

**Comment by simonjohansson** [comment]
@jtuchscherer 

Here is a breakdown on how many events we get with noaa.

![click to enter plot title](https://cloud.githubusercontent.com/assets/736248/9404945/f94ea1da-47f3-11e5-9f55-7a3639a3ddeb.png)

Doppler machines
```
nsaadmin@2027f501-9184-434f-9d94-0bfea08a8280:/var/vcap/sys/log$ find /var/vcap/sys/log -type f | xargs grep "Log message output too high" | wc -l
0
nsaadmin@310504fc-0638-42c1-925f-5d347f16555c:/var/vcap/sys/log$ find /var/vcap/sys/log -type f | xargs grep "Log message output too high" | wc -l
0
```

loggregator_trafficcontroller machines
```
nsaadmin@2b0fe1db-6a5e-4a16-a866-16b860ad4d3f:/var/vcap/sys/log$ find /var/vcap/sys/log -type f | xargs grep "Log message output too high" | wc -l
0
nsaadmin@2a79471f-63b0-4b55-b2bf-d5fcade54ef3:/var/vcap/sys/log$ find /var/vcap/sys/log -type f | xargs grep "Log message output too high" | wc -l
0
``` 

Let me know if you need any further info from my env.

**Comment by jtuchscherer** [comment]
@simonjohansson Thanks for providing the additional data. We are currently trying to reproduce this in our environments. 

Also, could you grep for the "Log message output too high" string not in the component logs, but in the output of your firehose. It is actually a log message sent by doppler to the client. See here (https://github.com/cloudfoundry/loggregator/blob/develop/src/doppler/truncatingbuffer/truncating_buffer.go#L87)

**Comment by simonjohansson** [comment]
@jtuchscherer 

> Also, could you grep for the "Log message output too high" string not in the component logs, but in the output of your firehose. It is actually a log message sent by doppler to the client. See here (https://github.com/cloudfoundry/loggregator/blob/develop/src/doppler/truncatingbuffer/truncating_buffer.go#L87)

Ah, ok, sorry, misunderstood there.
Will do, send you an update about this on monday.

**Comment by dsabeti** [comment]
Hi @simonjohansson,

We've tried a couple times to reproduce this on our environments, using the same app, without a lot of success. First we tried bosh-lite, but we could see all messages coming through. 

Then we tried a medium-sized AWS deployment, with a decent amount of traffic. First, we ran the following in the firehose_sample app:
```
DOPPLER_ADDR=[redacted] CF_ACCESS_TOKEN=[redacted] go run main.go | grep LogMessage | egrep "[app-guid]|Log message output too high" > firehose.txt
```
The idea was to save all the log messages associated with the app given by `app-guid`. We actually saw that we got 3 fewer messages than expected: if we made 100 requests and expected 200 log messages, we got 197 log messages, and it seemed always to drop the last 3 messages.

But then we tried a different command:
```
DOPPLER_ADDR=[redacted] CF_ACCESS_TOKEN=[redacted] go run main.go  > firehose.txt
cat firehose.txt | grep LogMessage | egrep "[app-guid]|Log message output too high"
```

The subtle difference here is that we saved all the firehose output first, then afterwards grepped for those log messages that we cared about. In this case, we saw the correct number of messages (200 log messages for 100 http requests).

We tried the same experiment scaling up to 2000 http requests, and noticed the exact same behavior. Ultimately, it seems like we're able to get all the log messages out of the firehose.

As a result, we think the problem has to do with your deployment, environment, or infrastructure. If you find any other information (such as the results of @jtuchscherer 's suggested experiment), please let us know and we'll follow up.

**Comment by simonjohansson** [comment]
@jtuchscherer
> Also, could you grep for the "Log message output too high"

No entries in my logs from the firehose.

> As a result, we think the problem has to do with your deployment, environment, or infrastructure. If you find any other information (such as the results of @jtuchscherer 's suggested experiment), please let us know and we'll follow up.

Ok great. Thanks for looking into this!

May I ask what the size of your VMs are?

I will close this for now and reopen in case I find something out.

**Comment by simonjohansson** [comment]
@jtuchscherer @dsabeti 

Some follow up, Im not 100% convinced this is due to our infra, but Im trying my hardest to nail it down to that.

What version were you trying your tests on? Im still on 212.

I have shut down all jobs in Z2, so when we curl a app the app lives on a Z1 DEA, gets routed by a Z1 router, logs gets forwarded to a Z1 doppler.
Only one app is running in the entire environment.

* 1 DEA
* 1 router
* 1 doppler

I conduct two tests
* Full blast 2k requests as fast as possible trough JMeter
* 100 curl and sleep 1 sec inbetween

And I see similar behaviour in both test cases.

```
$ ssh doppler_z1/0
$ hostname -I
10.230.18.100
$ cat run
export DOPPLER_ADDR=ws://10.230.18.100:8081
export CF_ACCESS_TOKEN=`cat ~/.cf/config.json | jq '.AccessToken' | xargs echo`
./main
$ sudo tcpdump -w /tmp/tcpdump
$ ./run > /tmp/firehose
$ # Fire off 2k requests as fast as possible, takes around 20 sec from localhost to CF.
$ egrep -a "App|RTR" /tmp/tcpdump | wc -l
4000
$  egrep -a "App|RTR" /tmp/firehose | wc -l
2860
$
$ sudo tcpdump -w /tmp/tcpdump
$ ./run > /tmp/firehose
$ # Fire off 100 slow requests, 1 sec sleep between each curl, takes around 100 seconds
$ egrep -a "App|RTR" /tmp/tcpdump | wc -l
200
$ egrep -a "App|RTR" /tmp/firehose | wc -l
144
```

As you can see in both cases I get the correct number of events from the DEA/Router with tcpdump, but only a subset of them reach the example firehose app, even though its configured to talk to doppler on localhost(I cannot connect to localhost as Doppler only binds to the pulic ip, but that should not make any difference in this case)

I will try with 215 and see if the behaviour is the same.

**Comment by simonjohansson** [comment]
Ok, 215 doesnt have the issue. 
Closing, and thanks again for having a look in your env!

**Comment by roxtar** [comment]
Just for completeness I believe this was the [story](https://www.pivotaltracker.com/story/show/97680512) which fixed the bug.

**Comment by simonjohansson** [comment]
Thanks @roxtar !
